### ISSUE TITLE
Replace scalar for-loops with existing IEngine hardware-accelerated tensor operations across codebase

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet#910]
## Overview

Comprehensive audit of the AiDotNet codebase reveals **200+ files with scalar for-loops that should use existing IEngine hardware-accelerated operations** from AiDotNet.Tensors. These operations already exist but are not being called — the codebase uses manual element-wise loops instead, missing out on SIMD/GPU acceleration.

This issue tracks the AiDotNet-side adoption work. The companion issue for **new** IEngine operations needed in AiDotNet.Tensors is tracked at: https://github.com/ooples/AiDotNet.Tensors/issues/16

### Key Metrics

| Metric | Count |
|--------|-------|
| Total for-loops in src/ | **10,094** |
| Current Engine.* calls (all ops) | **8,065** |
| Current Engine.Tensor* calls | **4,622** |
| Files with triple-nested for-loops (i,j,k) | **316** |
| Files with scalar AddTensors that should use TensorAdd | **41** |
| Files with scalar Concatenate that should use TensorConcatenate | **44** |
| Manual softmax implementations (should use Engine.Softmax) | **37** |
| Manual attention implementations (should use ScaledDotProductAttention) | **27+** |
| Manual LayerNorm implementations (should use Engine.LayerNorm) | **23** |

---

## Phase 1: Replace Scalar Arithmetic with Existing Tensor Ops (HIGH PRIORITY)

These IEngine operations **already exist** and can be adopted immediately.

### 1.1 TensorAdd — 41 files with scalar AddTensors loops

**Existing op**: `Engine.TensorAdd<T>(a, b)`

**Pattern to replace**:
```csharp
// BEFORE: Scalar loop
var resultSpan = result.AsWritableSpan();
var aSpan = a.AsSpan();
var bSpan = b.AsSpan();
for (int i = 0; i < resultSpan.Length; i++)
    resultSpan[i] = NumOps.Add(aSpan[i], bSpan[i]);

// AFTER: Hardware-accelerated
var result = Engine.TensorAdd<T>(a, b);
```

**Top priority files**:
- NeuralNetworks/Layers/ (residual connections in every layer)
- Finance/ (TimesFM, DeepState, NBEATS, TCN)
- Video/ (RAFT, FILM, RIFE optical flow)

### 1.2 TensorSubtract — 23 files with scalar loops

**Existing op**: `Engine.TensorSubtract<T>(a, b)`

**Pattern to replace**:
```csharp
// BEFORE
for (int i = 0; i < span.Length; i++)
    resultSpan[i] = NumOps.Subtract(aSpan[i], bSpan[i]);

// AFTER
var result = Engine.TensorSubtract<T>(a, b);
```

### 1.3 TensorMultiply — 13 files with scalar element-wise loops

**Existing op**: `Engine.TensorMultiply<T>(a, b)`

**Pattern to replace**:
```csharp
// BEFORE
for (int i = 0; i < span.Length; i++)
    resultSpan[i] = NumOps.Multiply(aSpan[i], bSpan[i]);

// AFTER
var result = Engine.TensorMultiply<T>(a, b);
```

### 1.4 TensorMultiplyScalar — 12+ files with scalar broadcast loops

**Existing op**: `Engine.TensorMultiplyScalar<T>(tensor, scalar)`

**Pattern to replace**:
```csharp
// BEFORE
var scale = NumOps.FromDouble(0.5);
for (int i = 0; i < span.Length; i++)
    resultSpan[i] = NumOps.Multiply(scale, span[i]);

// AFTER
var scale = NumOps.FromDouble(0.5);
var result = Engine.TensorMultiplyScalar<T>(tensor, scale);
```

### 1.5 TensorConcatenate — 44 files with scalar copy loops

**Existing op**: `Engine.TensorConcatenate<T>(tensors, axis)`

**Pattern to replace**:
```csharp
// BEFORE: Nested copy loops
var result = new Tensor<T>(combinedShape);
int offset = 0;
for (int t = 0; t < tensors.Length; t++)
{
    var src = tensors[t].AsSpan();
    var dst = result.AsWritableSpan();
    for (int i = 0; i < src.Length; i++)
        dst[offset + i] = src[i];
    offset += src.Length;
}

// AFTER
var result = Engine.TensorConcatenate<T>(tensors, axis: 0);
```

### 1.6 TensorExp — 2 files with scalar exp loops

**Existing op**: `Engine.TensorExp<T>(tensor)`

### 1.7 TensorPermute — 9 files with manual transpose loops

**Existing op**: `Engine.TensorPermute<T>(tensor, axes)`

**Pattern to replace**:
```csharp
// BEFORE: Triple-nested copy loops for transpose
for (int i = 0; i < dim0; i++)
    for (int j = 0; j < dim1; j++)
        for (int k = 0; k < dim2; k++)
            result[i, k, j] = input[i, j, k];

// AFTER
var result = Engine.TensorPermute<T>(input, new[] { 0, 2, 1 });
```

---

## Phase 2: Replace Manual Composite Operations (MEDIUM-HIGH PRIORITY)

These operations exist at the matrix level and can be used on tensor slices, or the existing higher-level IEngine ops can be called.

### 2.1 ScaledDotProductAttention — 27+ files with manual implementation

**Existing op**: `Engine.ScaledDotProductAttention<T>(Q, K, V, mask, scale)`

**Pattern to replace**:
```csharp
// BEFORE: 30+ lines of manual Q*K^T/sqrt(d) + softmax + V
var scores = Engine.TensorBatchMatMul<T>(Q, K_transposed);
// manual scale loop
// manual softmax loop (exp, sum, divide)
// manual mask application
var output = Engine.TensorBatchMatMul<T>(scores, V);

// AFTER: Single call
var output = Engine.ScaledDotProductAttention<T>(Q, K, V, mask, scale);
```

**Top priority files** (these run in the innermost loop of every transformer):
- NeuralNetworks/Layers/MultiHeadAttentionLayer.cs
- NeuralNetworks/Layers/SelfAttentionLayer.cs
- ComputerVision/SwinTransformer (window attention)
- ComputerVision/DETR, DINO, RTDETR (cross-attention)
- Video/SAM2, RAFT (temporal attention)
- Finance/Autoformer, TimesFM (temporal attention)

### 2.2 Softmax (Matrix-level) — 37 files with manual exp-sum-divide loops

**Existing op**: `Engine.Softmax(matrix)` (matrix-level)

**Pattern to replace**:
```csharp
// BEFORE: Manual numerically-stable softmax
var maxVal = span[0];
for (int i = 1; i < span.Length; i++)
    if (NumOps.GreaterThan(span[i], maxVal)) maxVal = span[i];
T sum = NumOps.Zero;
for (int i = 0; i < span.Length; i++)
{
    span[i] = NumOps.Exp(NumOps.Subtract(span[i], maxVal));
    sum = NumOps.Add(sum, span[i]);
}
for (int i = 0; i < span.Length; i++)
    span[i] = NumOps.Divide(span[i], sum);

// AFTER: Single call (on matrix slice or reshaped tensor)
var result = Engine.Softmax(matrix);
```

### 2.3 LayerNorm (Matrix-level) — 23 files with manual mean-variance-normalize

**Existing op**: `Engine.LayerNorm(input, gamma, beta, epsilon)`

**Pattern to replace**:
```csharp
// BEFORE: Manual mean, variance, normalize, affine transform
T mean = NumOps.Zero;
for (int i = 0; i < n; i++) mean = NumOps.Add(mean, span[i]);
mean = NumOps.Divide(mean, NumOps.FromDouble(n));
T variance = NumOps.Zero;
for (int i = 0; i < n; i++)
{
    var diff = NumOps.Subtract(span[i], mean);
    variance = NumOps.Add(variance, NumOps.Multiply(diff, diff));
}
// ... normalize and apply gamma/beta

// AFTER
var result = Engine.LayerNorm(input, gamma, beta, epsilon);
```

---

## Phase 3: Future Adoption (Blocked on AiDotNet.Tensors#16)

These replacements require new tensor-level IEngine operations that don't exist yet. They are tracked in https://github.com/ooples/AiDotNet.Tensors/issues/16 and can be adopted here once implemented.

| Future IEngine Op | Files waiting | Blocked on |
|-------------------|---------------|------------|
| TensorSigmoid | 14 | Tensors#16 Category 1.1 |
| TensorReLU | 4 | Tensors#16 Category 1.2 |
| TensorGELU | 1 | Tensors#16 Category 1.3 |
| TensorSiLU (Swish) | 3 | Tensors#16 Category 1.4 |
| TensorTanh | 1 | Tensors#16 Category 1.5 |
| TensorArgMax (CPU) | 45 | Tensors#16 Category 2.1 |
| TensorArgMin | ~10 | Tensors#16 Category 2.2 |
| TensorMinValue | completeness | Tensors#16 Category 2.3 |
| TensorReduceMean (axis) | 4 | Tensors#16 Category 2.4 |
| TensorVariance / TensorStd | 12 | Tensors#16 Category 2.5 |
| TensorSoftmax (tensor-level) | 37 | Tensors#16 Category 3.1 |
| TensorLayerNorm (tensor-level) | 23 | Tensors#16 Category 3.2 |
| TensorInterpolate (Lerp) | 11 | Tensors#16 Category 3.3 |
| TensorAddNoise (fused) | 8 | Tensors#16 Category 3.4 |
| TensorSplit | 65 | Tensors#16 Category 3.5 |
| TensorMaxPool2D | 26 | Tensors#16 Category 4.1 |
| TensorAvgPool2D | 26 | Tensors#16 Category 4.2 |
| TensorConv2D | 18 | Tensors#16 Category 4.3 |

---

### Priority 1: NeuralNetworks/Layers/ (26 files)

Core layer implementations used by ALL neural networks. Fixing these provides the widest impact.

**Key files**:
- MultiHeadAttentionLayer.cs — manual attention, softmax, concatenation
- SelfAttentionLayer.cs — manual attention, softmax
- ConvolutionalLayer.cs — manual conv, pooling loops
- BatchNormalizationLayer.cs — manual mean/variance computation
- LayerNormalizationLayer.cs — manual layer norm
- ResidualBlock layers — manual tensor addition for skip connections
- All activation layers — scalar activation function calls

### Priority 2: Video/ (14 files)

- RAFT/ — optical flow with manual attention and conv
- FILM/ — frame interpolation with manual interpolation loops
- RIFE/ — real-time interpolation with manual tensor ops
- SAM2/ — segment anything with manual attention
- ActionRecognition/ — classification heads with manual argmax

### Priority 3: Finance/ (11 files)

- TimesFM/ — temporal transformer with manual attention
- DeepState/ — state space model with manual tensor ops
- NBEATS/ — basis expansion with manual addition
- TCN/ — temporal convolution with manual conv loops
- WaveNet/ — dilated convolution with manual implementation
- Autoformer/ — auto-correlation with manual softmax

### Priority 4: ComputerVision/ (8 files)

- SwinTransformer/ — window attention with manual softmax
- DETR, DINO, RTDETR/ — detection transformers with manual cross-attention
- YOLOv9/ — detection with manual conv and pooling
- TrOCR/ — OCR transformer with manual attention

### Priority 5: Audio/ (4 files)

- ECAPA-TDNN/ — speaker verification with manual attention
- VoxLingua107/ — language identification with manual pooling

### Priority 6: Diffusion/ (partially done)

The diffusion module has been partially converted in PR #897. Remaining scalar patterns in diffusion are mostly conditional/activation patterns that need new tensor-level ops (blocked on Tensors#16).

---

### Pattern for classes inheriting from base with Engine property
```csharp
// Classes inheriting from DiffusionModelBase, LayerBase, NoisePredictorBase, etc.
// have `protected IEngine Engine => AiDotNetEngine.Current;`
var result = Engine.TensorAdd<T>(a, b);
```

### Pattern for standalone classes
```csharp
// Standalone classes (not inheriting from a base with Engine)
using AiDotNet.Engines;

var engine = AiDotNetEngine.Current;
var result = engine.TensorAdd<T>(a, b);
```

### Common replacement patterns

**Noise mixing** (diffusion forward process):
```csharp
var scaled1 = Engine.TensorMultiplyScalar<T>(signal, alpha);
var scaled2 = Engine.TensorMultiplyScalar<T>(noise, sigma);
return Engine.TensorAdd<T>(scaled1, scaled2);
```

**CFG guidance**:
```csharp
var diff = Engine.TensorSubtract<T>(condPred, uncondPred);
var scaled = Engine.TensorMultiplyScalar<T>(diff, scaleT);
return Engine.TensorAdd<T>(uncondPred, scaled);
```

**Linear interpolation**:
```csharp
var s1 = Engine.TensorMultiplyScalar<T>(a, NumOps.FromDouble(1.0 - t));
var s2 = Engine.TensorMultiplyScalar<T>(b, NumOps.FromDouble(t));
return Engine.TensorAdd<T>(s1, s2);
```

---

## Summary

| Phase | Work | Files | Can Start Now? |
|-------|------|-------|----------------|
| Phase 1 | Replace scalar arithmetic with existing TensorAdd/Sub/Mul/etc. | ~133 | Yes |
| Phase 2 | Replace manual Softmax/LayerNorm/Attention with existing ops | ~87 | Yes |
| Phase 3 | Adopt new tensor-level ops once AiDotNet.Tensors adds them | ~200+ | Blocked on Tensors#16 |

**Total estimated impact**: 200+ files across every major module (NeuralNetworks, Diffusion, Video, Finance, ComputerVision, Audio, TimeSeries) will benefit from hardware-accelerated tensor operations instead of scalar for-loops.

**Related**: https://github.com/ooples/AiDotNet.Tensors/issues/16 (new IEngine operations needed)
